plutus-example/plutus-example/src/Cardano/PlutusExample/ScriptContextChecker.hs

```diff
@@ -0,0 +1,372 @@
+  :: UTxO era
+  -> EpochInfo (Either TransactionValidityIntervalError)
+  -> SystemStart
+  -> Api.Tx era
+  -> Either ScriptContextError MyCustomRedeemer
+txToCustomRedeemer _ _ _ (ByronTx _) = Left NoScriptsInByronEra
+txToCustomRedeemer utxo eInfo sStart (ShelleyTx ShelleyBasedEraAlonzo ledgerTx) = do
+  let txBody = Alonzo.body ledgerTx
+      witness = Alonzo.wits ledgerTx
+      mTxIns = Prelude.map (Alonzo.txInfoIn (toLedgerUTxO ShelleyBasedEraAlonzo utxo)) . Set.toList $ Alonzo.inputs txBody
+      mTouts = Prelude.map Alonzo.txInfoOut $ seqToList $ Alonzo.outputs txBody
+      minted = Alonzo.transValue $ Alonzo.mint txBody
+      txfee = Alonzo.transValue . toMaryValue . lovelaceToValue . fromShelleyLovelace $ Alonzo.txfee txBody
+      Alonzo.TxDats datumHashMap = Alonzo.txdats witness
+      datumHashes = Prelude.map Alonzo.transDataPair $ Map.toList datumHashMap
+      _txid = Alonzo.txInfoId . toShelleyTxId $ getTxIdShelley ShelleyBasedEraAlonzo txBody
+      txcerts = Prelude.map Alonzo.transDCert . seqToList $ Alonzo.txcerts txBody
+      txsignatories = Prelude.map Alonzo.transKeyHash . Set.toList $ Alonzo.reqSignerHashes txBody
+  valRange <-
+    first IntervalConvError
+      $ Alonzo.transVITime eInfo sStart $ Alonzo.txvldt txBody
+
+  tOuts <- if Prelude.all M.isJust mTouts
+           then return $ catMaybes mTouts
+           else Prelude.error "Tx Outs not all Just"
+  txins <- if Prelude.all M.isJust mTxIns
+           then return $ catMaybes mTxIns
+           else Prelude.error "Tx Ins not all Just"
+  Right $ MyCustomRedeemer tOuts txins minted valRange txfee datumHashes txcerts txsignatories
+ where
+  seqToList (x Seq.:<| rest) = x : seqToList rest
+  seqToList Seq.Empty = []
+
+txToCustomRedeemer _ _ _ (ShelleyTx _ _) = Left NoScriptsInEra
+
+testScriptContextToScriptData :: MyCustomRedeemer -> ScriptData
+testScriptContextToScriptData = fromPlutusData . PlutusTx.builtinDataToData . toBuiltinData
+
+readCustomRedeemerFromTx
+  :: FilePath
+  -> AnyConsensusModeParams
+  -> NetworkId
+  -> ExceptT ScriptContextError IO MyCustomRedeemer
+readCustomRedeemerFromTx fp (AnyConsensusModeParams cModeParams) network = do
+  InAnyCardanoEra cEra alonzoTx
+    <- firstExceptT ReadTxBodyError
+         . newExceptT
+         $ readFileTextEnvelopeAnyOf
+             [ FromSomeType (AsTx AsAlonzoEra) (InAnyCardanoEra AlonzoEra)
+             ]
+             fp
+
+  sbe <- getSbe $ cardanoEraStyle cEra
+  SocketPath sockPath <- firstExceptT EnvVarSocketErr readEnvSocketPath
+  case consensusModeOnly cModeParams of
+    CardanoMode -> do
+      let localNodeConnInfo = LocalNodeConnectInfo cModeParams network sockPath
+
+      (_chainTip, eQueryTipLocalState) <- liftIO $ queryQueryTip localNodeConnInfo Nothing
+      qTipLocalState <- firstExceptT AcquireFail $ hoistEither eQueryTipLocalState
+      let eInfo = epochInfo qTipLocalState
+      sStart <- hoistMaybe NoSystemStartTimeError $ mSystemStart qTipLocalState
```

Comment:
`queryQueryTip` also queries for era:

```
era <- queryExpr (QueryCurrentEra CardanoModeIsMultiEra)
```

Which is thrown away.  Each `queryExpr` is a roundtrip to the server.  `_chainTip` is also discarded.

Generally speaking it's better to not query for things you don't need to avoid latency and placing load on the node if we don't need to.

It's better to inline `queryQueryTip`, replace the call to `executeLocalStateQueryExprWithChainSync` with `executeLocalStateQueryExpr` and and drop the call to `queryExpr (QueryCurrentEra CardanoModeIsMultiEra)` do avoid querying for things you don't need.